src/mame/mit/cadr_iob.cpp

```diff
@@ -0,0 +1,644 @@
+//#define LOG_OUTPUT_FUNC osd_printf_info
+#include "logmacro.h"
+
+
+
+DEFINE_DEVICE_TYPE(CADR_IOB, cadr_iob_device, "cadr_iob", "CADR I/O Board")
+
+
+cadr_iob_device::cadr_iob_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, CADR_IOB, tag, owner, clock)
+	, m_i8748(*this, "i8748")
+	, m_irq_vector_cb(*this)
+	, m_keyboard(*this, "KEY.%u", 0)
+	, m_mouse_x(*this, "MOUSE_X")
+	, m_mouse_y(*this, "MOUSE_Y")
+	, m_mouse_buttons(*this, "MOUSE_BUTTONS")
+	, m_my_chaos_address(*this, "MY_CHAOS_ADDRESS")
+	, m_speaker(*this, "speaker")
+{
+}
+
+
+void cadr_iob_device::device_add_mconfig(machine_config &config)
+{
+	I8748(config, m_i8748, 6_MHz_XTAL);
+	m_i8748->bus_in_cb().set(FUNC(cadr_iob_device::mcu_bus_r));
+	m_i8748->bus_out_cb().set(FUNC(cadr_iob_device::mcu_bus_w));
+	m_i8748->p1_out_cb().set(FUNC(cadr_iob_device::mcu_p1_w));
+	m_i8748->p2_in_cb().set(FUNC(cadr_iob_device::mcu_p2_r));
+
+	SPEAKER(config, "mono").front_center();
+	SPEAKER_SOUND(config, m_speaker).add_route(ALL_OUTPUTS, "mono", 1.00);
+}
+
+
+ROM_START(cadr_keyboard)
+	ROM_REGION(0x400, "i8748", 0)
+	ROM_LOAD("cadr_kb_8748_3-7-83.bin", 0x0, 0x400, CRC(8467ad0e) SHA1(cefbbd8f0afb86bd0157041f868fa70b38848582))
+ROM_END
+
+
+const tiny_rom_entry *cadr_iob_device::device_rom_region() const
+{
+	return ROM_NAME(cadr_keyboard);
+}
+
+
+static INPUT_PORTS_START(keyboard)
+	PORT_START("KEY.0")
+	PORT_BIT(0x01, IP_ACTIVE_HIGH, IPT_UNKNOWN) // f2 02 00
+	PORT_BIT(0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_NAME("Ⅱ")// f2 02 02
+	PORT_BIT(0x04, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_NAME("Ⅳ") // f2 02 04
+	PORT_BIT(0x08, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_NAME("MODE LOCK") // f2 02 06
+	PORT_BIT(0x10, IP_ACTIVE_HIGH, IPT_UNKNOWN) // f2 02 08
+	PORT_BIT(0x20, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_NAME("L.SUPER") // f2 02 0a
+	PORT_BIT(0x40, IP_ACTIVE_HIGH, IPT_UNKNOWN) // f2 02 0c
+	PORT_BIT(0x80, IP_ACTIVE_HIGH, IPT_UNKNOWN) // f2 02 0e
+
+	PORT_START("KEY.1")
+	PORT_BIT(0x01, IP_ACTIVE_HIGH, IPT_UNKNOWN) // f2 02 10
+	PORT_BIT(0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_4) PORT_CHAR('4') PORT_CHAR('$') // 4 f2 02 12
+	PORT_BIT(0x04, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_R) PORT_CHAR('r') PORT_CHAR('R') // R f2 02 14
+	PORT_BIT(0x08, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_F) PORT_CHAR('f') PORT_CHAR('F') // F f2 02 16
+	PORT_BIT(0x10, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_V) PORT_CHAR('v') PORT_CHAR('V') // V f2 02 18
+	PORT_BIT(0x20, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_NAME("ALT LOCK") // f2 02 1a
+	PORT_BIT(0x40, IP_ACTIVE_HIGH, IPT_UNKNOWN) // f2 02 1c
+	PORT_BIT(0x80, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_RIGHT) PORT_CHAR(UCHAR_MAMEKEY(RIGHT)) // f2 02 1e
+
+	PORT_START("KEY.2")
+	PORT_BIT(0x01, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_LCONTROL) PORT_CHAR(UCHAR_SHIFT_2) // f2 02 20
+	PORT_BIT(0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_NAME(": ±") // f2 02 22
```

Comment:
Doe this actually produce colon and ± characters, and if so, is there a good reason not to use `PORT_CHAR` for this?